File: src/api/subsonic/methods/getArtist.ts
import * as v from 'valibot';

import {
  handleJsonResponse,
  makeSubsonicRequest,
  parseJsonResponse,
} from '../makeRequest';
import { BaseAlbum } from '../types/baseAlbum';
import { BaseArtist } from '../types/baseArtist';

export function subsonicGetArtist(id: string) {
  return makeSubsonicRequest({ method: 'rest/getArtist', id })
    .flatMap(response =>
      parseJsonResponse(response, {
        artist: v.object({ ...BaseArtist.entries, album: v.array(BaseAlbum) }),
      }),
    )
    .flatMap(json => handleJsonResponse(json['subsonic-response']))
    .map(result => result.artist);
}

File: src/api/subsonic/getArtist.ml
open Fx
open Monad_syntax
open Utils

let make id =
  let* { response = { status = `ok; artist } } =
    let params = Js.Dict.empty () in
    Js.Dict.set params "id" (Value id);
    makeRequest { _method = "rest/getArtist"; params } Artist.of_json
  in
  Ok artist
